✨ [bento][amp-accordion] CSS Port draft for amp accordion (PR #2)

krdwan opened the pull request:
Partial #30445.

Having a problem with rebasing on the previous PR (https://github.com/ampproject/amphtml/pull/30593), so created a new PR instead.

Porting CSS over from `0.1`.

- Added styling for bento and amp modes for `header` and `content` sections.
- Preact mode is controlled by JSS
- amp mode is controlled by component css styling

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ampproject/amphtml?pullRequest=30622) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/ampproject/amphtml?pullRequest=30622) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ampproject/amphtml?pullRequest=30622) it.</sub>

dvoytenko commented on extensions/amp-accordion/1.0/accordion.js:
```diff
@@ -48,6 +48,13 @@ const CHILD_STYLE = {
 
+const SECTION_HEADER_STYLE = {
+  cursor: 'pointer',
+  backgroundColor: '#efefef',
```
Cursor pointer is ok here. But we can't put any of the rest into the inline style - this would make them un-overridable in stylesheets. Let's remove these styles for now. If the wg decides we want these styles here, we will have to bring JSS back. But for now it's easier to not have it, imho.

dvoytenko commented on extensions/amp-accordion/1.0/accordion.js:
```diff
@@ -152,6 +159,8 @@ export function AccordionSection({
   contentAs: ContentComp = 'div',
   expanded: defaultExpanded = false,
   animate: defaultAnimate = false,
+  headerStyle = {},
```
No need for defaults here.

krdwan replied:
Thanks, removed.

dvoytenko commented on extensions/amp-accordion/1.0/amp-accordion.css:
```diff
@@ -13,3 +13,19 @@
+}
+
+/* heading
+ * TODO(dvoytenko): remove most of these styles, except maybe for `cursor`.
```
Rewrite TODO as `TODO(#30445): ...`.

krdwan replied:
Updated.

dvoytenko commented on extensions/amp-accordion/1.0/amp-accordion.js:
```diff
@@ -128,6 +129,7 @@ const bindSectionShimToElement = (element) => SectionShim.bind(null, element);
  */
 function HeaderShim(sectionElement, {onClick}) {
   const headerElement = sectionElement.firstElementChild;
+  headerElement.classList.add('i-amphtml-accordion-header');
```
This is a side-effect. Let's move it to the `useLayoutEffect` below. Alternatively, we can move this code to `getState`.

krdwan replied:
Updated.

dvoytenko commented on extensions/amp-accordion/1.0/amp-accordion.js:
```diff
@@ -157,6 +159,7 @@ const bindHeaderShimToElement = (element) => HeaderShim.bind(null, element);
  */
 function ContentShimWithRef(sectionElement, {hidden}, ref) {
   const contentElement = sectionElement.lastElementChild;
+  contentElement.classList.add('i-amphtml-accordion-content');
```
Ditto.

krdwan replied:
Updated.

krdwan commented:
Latest revision includes:

1. Moved styling in Preact to use JSX per team discussion.
2. This includes `CHILD_STYLING`(CSS related to animations)
3. Addressed review comments.

dvoytenko commented on extensions/amp-accordion/1.0/accordion.js:
```diff
@@ -37,17 +38,6 @@ const EMPTY_EXPANDED_MAP = {};
 
 const generateSectionId = sequentialIdGenerator();
 
-const CHILD_STYLE = {
```
These particular styles are not overridable. As such, we can continue to inline them. If you still want them in the JSS - it's totally reasonable, but we need to add `!important`.

krdwan replied:
I personally prefer using JSS to keep all styling together.  I think it improves readability so the developer can quickly look into the one JSS file to see all styling.  I have added `!important`.  If there's something else I should consider, don't hesitate to let me know :).  Thanks!

dvoytenko reviewed (changes requested):
Looks good, but one request.

dvoytenko commented on extensions/amp-accordion/1.0/test/test-accordion.js:
```diff
@@ -119,18 +119,39 @@ describes.sandboxed('Accordion preact component', {}, (env) => {
+      expect(content2.hidden).to.be.true;
+
+      // Styling.
+      expect(header0.className.includes('section-child')).to.be.true;
```
IMHO instead of testing for class names (or with them) I'd like to attach the accordion to DOM and call `getComputedStyle()` to check critical styles - the ones we marked with `!important`. Otherwise, the class  presence does not guarantee that we didn't make a bug with the class definition itself.

krdwan replied:
Sounds good Dima.  For the `test-amp-accordion.js` (amp mode tests) I have made the updates.

For `test-accordion.js` (preact tests), I have not been able to get the tests to work.  A few things I tried include:

1. Attach the component DOM node to its own `ownerDocument.defaultView` (see lines 115 - 117).  However, I don't think the `style` tag from JSS is installed here, so not getting the styles as expected.

2. I also tried using options in the mount section (see line 109).  However, I'm unable to attach the component DOM node as `win.document` is apparently `undefined`.

I'm not too familiar with the scoping in the test environments.  Do you have any recommendations on where I can find out more about this?

caroqliu replied:
While these tests are super thorough and help increase confidence, it's not common to test CSS properties this granularly inside the components for visual styles. If it's not too much trouble, what do you think about copying the visual tests from 0.1 over? It should give us just as much information without as much of this tedium.

krdwan replied:
Thanks!  As discussed offline, I've added an item in the tracker issue to look into the visual tests separately from this PR. https://github.com/ampproject/amphtml/issues/30445

dvoytenko reviewed (approved):
LGTM with one ask for tests.

krdwan commented on extensions/amp-accordion/1.0/accordion.js:
```diff
@@ -152,6 +142,8 @@ export function AccordionSection({
   expanded: defaultExpanded = false,
   animate: defaultAnimate = false,
+  headerClassName,
+  contentClassName,
```
I'm fine with this, updated.

krdwan commented on extensions/amp-accordion/1.0/accordion.jss.js:
```diff
@@ -0,0 +1,50 @@
+};
+
+// TODO(#30445): update these styles after team agrees on styling
+// or delete is not used
```
Updated!
